pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/asm/ParseLock.java

```diff
@@ -32,14 +32,22 @@ private ParseStatus getFinalStatus() {
                     try {
                         boolean success = doParse();
                         status = success ? ParseStatus.FULL : ParseStatus.FAILED;
-                        this.status = status;
                         finishParse(!success);
                     } catch (Throwable t) {
                         status = ParseStatus.FAILED;
-                        this.status = status;
                         LOG.error(t.toString(), t);
                         finishParse(true);
                     }
+
+                    // the status must be updated as last statement, so that
```

Comment:
I think, this is the reason why other threads might see ClassStub.enclosingInfo as null - it is set by finishParse() (at least for simple top level classes), but finishParse() was called after the status was set FULL (finished). That means, another thread wouldn't have waited until finishParse() is finished... I could reproduce this problem by adding a Thread.yield() or a Thread.sleep() at the beginning of finishParse(), just before enclosingInfo is set.

Reply:
Update: a Thread.sleep() works better to reproduce the exceptions.